content/blog/waddlesplash/2023-04-24_condition_variables.md

```diff
@@ -0,0 +1,416 @@
++++
+type = "blog"
+title = "Haiku's (Kernel) Condition Variables API: Design & Implementation"
+author = "waddlesplash"
+date = "2023-04-24 12:00:00-04:00"
+tags = ["kernel"]
++++
+
+A few months after my contract with Haiku, Inc. began, I rewrote the implementation of the Haiku kernel's condition variables. As this new implementation has run in Haiku for over a year and shipped in the latest release with no sign of any remaining issues, I figured it is high time for a deep-dive on the API, its implementation history, and the design of the new implementation I wrote.
+
+I expect this article will be of broader interest than just to Haiku’s community, because Haiku’s condition variables API has some notable (and powerful) features not found in those of other operating systems, and its implementation is thus likewise unique (at least, as far as I have been able to figure out.)
+
+<!--more-->
+
+## Background
+
+Before delving in to Haiku’s condition variables and how they differ from those of other operating systems, it is first important to understand what they are for, and how they fit in with the other thread-related APIs. (For those already very familiar with atomics and thread synchronization primitives, you can likely skip this section entirely, and jump directly to the section titled “Condition variables”.)
+
+Teams (or in more common terminology, “processes”) are built around *threads*, or more properly, *threads of execution*. Any team will have at least one thread (the “main thread”), and any number of other threads. On systems with multiple CPU cores (and/or “hyperthreads”), multiple threads will be running at once; and even on single-core systems, the order in which threads run will not always be the same, since it may depend on disk I/O, system load, etc. Thus, thread operations need to be *synchronized.*
+
+### Atomics
+
+Consider if I have a variable which I want to add 1 to.
+```cpp
+int32 gSharedInteger = /* initializer... */;
+void add() {
+	int32 localInteger = gSharedInteger;
+	localInteger = localInteger + 1;
+	gSharedInteger = localInteger;
+}
+```
+What happens if `add` is invoked for the first time simultaneously in two separate threads? Will `sharedInteger` become `2`, or (if the invocations really occur at exactly the same instant), will it become `1`, since the threads will have “raced” against each other?
+
```

Comment:
I would include a sequence diagram showing how this can go wrong, but maybe this is not the point of this article? Can we link to something that has additional details, a wikipedia page or so?

Reply:
Wikipedia has a diagram of sorts, indeed. I'll just link to it.